src/backend/catalog-builder/catalog-builder.lambda.ts

```diff
@@ -160,6 +129,74 @@ async function* relevantObjects(bucket: string) {
   } while (request.ContinuationToken != null);
 }
 
+async function appendPackage(packages: any, assemblyKey: string, bucketName: string) {
+  console.log(`Processing key: ${assemblyKey}`);
+  const pkgKey = assemblyKey.replace(constants.ASSEMBLY_KEY_SUFFIX, constants.PACKAGE_KEY_SUFFIX);
+  const [, packageName, versionStr] = pkgKey.match(KEY_FORMAT_REGEX)!;
+  const version = new SemVer(versionStr);
+  const found = packages.get(packageName)?.get(version.major);
+  if (found != null && version.compare(found.version) <= 0) {
+    console.log(`Skipping ${packageName}@${version} because it is not newer than the existing ${found.version}`);
+    return;
+  }
+  console.log(`Registering ${packageName}@${version}`);
+
+  // Donwload the tarball to inspect the `package.json` data therein.
+  const pkg = await aws.s3().getObject({ Bucket: bucketName, Key: pkgKey }).promise();
+  const metadataKey = pkgKey.replace(constants.PACKAGE_KEY_SUFFIX, constants.METADATA_KEY_SUFFIX);
+  const metadataResponse = await aws.s3().getObject({ Bucket: bucketName, Key: metadataKey }).promise();
+  const manifest = await new Promise<Buffer>((ok, ko) => {
+    gunzip(Buffer.from(pkg.Body!), (err, tar) => {
+      if (err) {
+        return ko(err);
+      }
+      extract()
+        .on('entry', (header, stream, next) => {
+          if (header.name !== 'package/package.json') {
+            // Not the file we are looking for, skip ahead (next run-loop tick).
+            return setImmediate(next);
+          }
+          const chunks = new Array<Buffer>();
+          return stream
+            .on('data', (chunk) => chunks.push(Buffer.from(chunk)))
+            .once('end', () => {
+              ok(Buffer.concat(chunks));
+              next();
+            })
+            .resume();
+        })
+        .once('finish', () => {
+          ko(new Error('Could not find package/package.json in tarball!'));
+        })
+        .write(tar, (writeErr) => {
+          if (writeErr) {
+            ko(writeErr);
+          }
+        });
+    });
+  });
+    // Add the PackageInfo into the working set
+  const pkgMetadata = JSON.parse(manifest.toString('utf-8'));
+  const npmMetadata = JSON.parse(metadataResponse?.Body?.toString('utf-8') ?? '{}');
+  const major = new SemVer(pkgMetadata.version).major;
+  if (!packages.has(pkgMetadata.name)) {
+    packages.set(pkgMetadata.name, new Map());
+  }
+  packages.get(pkgMetadata.name)!.set(major, {
```

Comment:
```suggestion
  // Add the PackageInfo into the working set
  const pkgMetadata = JSON.parse(manifest.toString('utf-8'));
  const npmMetadata = JSON.parse(metadataResponse?.Body?.toString('utf-8') ?? '{}');
  const major = new SemVer(pkgMetadata.version).major;
  if (!packages.has(pkgMetadata.name)) {
    packages.set(pkgMetadata.name, new Map());
  }
  packages.get(pkgMetadata.name)!.set(major, {
```

---

What if this package version is _older_ than the current tip of this major version? This can happen for example if someone publishes `2.3.4` after `2.4.0` has been published. We probably don't want to roll the catalog back to an older revision.

Reply:
@RomainMuller I haven't changed that logic...isn't this handled here:

https://github.com/cdklabs/construct-hub/blob/e9ccda58befba3a17562cb703ca9c20e33a5d399/src/backend/catalog-builder/catalog-builder.lambda.ts#L138-L142

Reply:
Oh right sorry!